contracts/LockedNORI.sol

```diff
@@ -40,127 +63,339 @@ contract LockedNORI is ERC777Upgradeable, ERC20PresetMinterPauserUpgradeable, ER
-    return unlockedBalanceOf(account);
+  /**
+   * @dev balanceOf: Total amount granted to the user less any previously claimed amount
+   */
+  function balanceOf(address account)
+    public
+    view
+    override(ERC20Upgradeable, ERC777Upgradeable)
+    returns (uint256)
+  {
+    return ERC777Upgradeable.balanceOf(account);
   }
 
-  function _beforeTokenTransfer(
-    address from,
-    address to,
-    uint256 amount
-  ) internal override(ERC20PresetMinterPauserUpgradeable, ERC20Upgradeable) {
-    if (to == address(0)) {
-      uint256 availableBalance = balanceOf(from);
-      uint256 wrappedBalance = super.balanceOf(from);
-      console.log(
-        "amount: %d  available: %d  wrapped: %d",
-        amount,
-        availableBalance,
-        wrappedBalance
-      );
-      require(amount <= availableBalance, "Withdrawl amount unavailable");
+  /**
+   * @dev getGrant: Returns all governing settings on a grant.
+   */
+  function getGrant(address account)
+    public
+    view
+    returns (
+      uint256 amount,
+      uint256 startTime,
+      uint256 vestEndTime,
+      uint256 unlockEndTime,
+      uint256 cliff1Time,
+      uint256 cliff2Time,
+      uint256 vestCliff1Amount,
+      uint256 vestCliff2Amount,
+      uint256 unlockCliff1Amount,
+      uint256 unlockCliff2Amount,
+      uint256 claimedAmount,
+      uint256 originalAmount
+    )
+  {
+    TokenGrant storage grant = grants[account];
+    return (
+      grant.grantAmount,
+      grant.lockupSchedule.startTime,
+      grant.vestingSchedule.endTime,
+      grant.lockupSchedule.endTime,
+      grant.lockupSchedule.cliffs[0].time,
+      grant.lockupSchedule.cliffs[1].time,
+      grant.vestingSchedule.cliffs[0].amount,
+      grant.vestingSchedule.cliffs[1].amount,
+      grant.lockupSchedule.cliffs[0].amount,
+      grant.lockupSchedule.cliffs[1].amount,
+      grant.claimedAmount,
+      grant.originalAmount
+    );
+  }
+
+  function _isTokenGranter(address account) internal view returns (bool) {
+    for (uint256 i = 0; i < getRoleMemberCount(TOKEN_GRANTER_ROLE); i++) {
+      if (getRoleMember(TOKEN_GRANTER_ROLE, i) == account) {
+        return true;
+      }
     }
-    super._beforeTokenTransfer(from, to, amount);
+    return false;
   }
 
-  function _afterTokenTransfer(
+  /**
+   * @dev called before send and transfer and used to disable transferring locket nori
+   */
+  function _beforeTokenTransfer(
+    address,
     address from,
     address to,
     uint256 amount
   ) internal override {
-    if (to == address(0)) {
-      TokenGrant storage grant = grants[from];
-      grant.claimedAmount += amount;
+    bool isNotMinting = from != address(0); // if it's not minting, then we check balances
+    // this is a burn / unwrap
+    if (isNotMinting) {
+      uint256 availableBalance = unlockedBalanceOf(from);
+      if(!hasRole(DEFAULT_ADMIN_ROLE, _msgSender())){
```

Comment:
I added this for revoking, but I wonder if it's the best approach

Reply:
I think we need it due to the operatorSend in `revokeUnvestedTokens` but should probably be tracking granter role rather than admin.  That also means all granters need to be operators I guess?

Reply:
I added an `it.todo` to test this.